doc/dev/custom_devices.rst

```diff
@@ -0,0 +1,118 @@
+            }
+            [[nodiscard]] virtual auto Zero() const -> Result { return NULL; }
+            [[nodiscard]] virtual auto One() const -> Result { return NULL; }
+            virtual auto Observable(ObsId, const std::vector<std::complex<double>> &,
+                                    const std::vector<QubitIdType> &) -> ObsIdType
+            {
+                return 0;
+            }
+            virtual auto TensorObservable(const std::vector<ObsIdType> &) -> ObsIdType { return 0; }
+            virtual auto HamiltonianObservable(const std::vector<double> &, const std::vector<ObsIdType> &)
+                -> ObsIdType
+            {
+                return 0;
+            }
+            virtual auto Measure(QubitIdType) -> Result
+            {
+                bool *ret = (bool *)malloc(sizeof(bool));
+                *ret = true;
+                return ret;
+            }
+
+            virtual void ReleaseQubit(QubitIdType) {}
+            virtual void ReleaseAllQubits() {}
+            [[nodiscard]] virtual auto GetNumQubits() const -> size_t { return 0; }
+            virtual void SetDeviceShots(size_t shots) {}
+            [[nodiscard]] virtual auto GetDeviceShots() const -> size_t { return 0; }
+            virtual void StartTapeRecording() {}
+            virtual void StopTapeRecording() {}
+            virtual void PrintState() {}
+            virtual void NamedOperation(const std::string &, const std::vector<double> &,
+                                        const std::vector<QubitIdType> &, bool)
+            {
+            }
+
+            virtual void MatrixOperation(const std::vector<std::complex<double>> &,
+                                         const std::vector<QubitIdType> &, bool)
+            {
+            }
+
+            virtual auto Expval(ObsIdType) -> double { return 0.0; }
+            virtual auto Var(ObsIdType) -> double { return 0.0; }
+            virtual void State(DataView<std::complex<double>, 1> &) {}
+            virtual void Probs(DataView<double, 1> &) {}
+            virtual void PartialProbs(DataView<double, 1> &, const std::vector<QubitIdType> &) {}
+            virtual void Sample(DataView<double, 2> &, size_t) {}
+            virtual void PartialSample(DataView<double, 2> &, const std::vector<QubitIdType> &, size_t) {}
+            virtual void Counts(DataView<double, 1> &, DataView<int64_t, 1> &, size_t) {}
+
+            virtual void PartialCounts(DataView<double, 1> &, DataView<int64_t, 1> &,
+                                       const std::vector<QubitIdType> &, size_t)
+            {
+            }
+
+            virtual void Gradient(std::vector<DataView<double, 1>> &, const std::vector<size_t> &) {}
+        };
+
+In addition to implementing the ``QuantumDevice`` class, one must implement the following method:
+
+.. code-block:: c++
+
+    extern "C" Catalyst::Runtime::QuantumDevice*
+    getCustomDevice() { return new CustomDevice(); }
```

Comment:
Here's the reference to this device constructor factory method: https://github.com/PennyLaneAI/catalyst/blob/38953d8c28b3b73d802b1cc10c6062e84874ed79/runtime/include/QuantumDevice.hpp#L297

We removed it in PR #89 in a favor of the multi-device support via passing their names from the frontend.

Reply:
I don't think it is problematic. Look here: https://github.com/PennyLaneAI/catalyst/pull/327/files#diff-4718a857bb4ed45b3ed299bdb7926e5beca237e53aa645d124f295e0cbbd2357R232

_device_ptr is a unique_ptr. We cannot use a `std::unique_ptr<QuantumDevice>` here because this is a C function, which does not know about C++ templates.

Reply:
Besides, this interface will likely change. We need to model __actual__ hardware device availability. How many devices are installed and whether they are currently being used by another process, etc, etc.

Reply:
I didn't reach this far when I had this comment! I agree the use of `unique_ptr` for `_driver_so_ptr` will indeed bound this dynamically allocated memory. However, I may reckon a better design in the future. :thinking: 

Also there shouldn't be any problem with using `std::unique_ptr` with C externed method. I'll double-check but I'm pretty sure it's possible.